lib/models/plugin-registry.js

```diff
@@ -0,0 +1,205 @@
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(config), addons, 'config.pipeline.runOrder');
+
+    return config;
+  },
+
+  _installedPlugins: function(addons, aliasMap) {
+    return Object.keys(addons)
+      .map(function(addonName) {
+        var addon   = addons[addonName];
+        var aliases = aliasMap[addonName].as;
+
+        return aliases.map(function(alias) {
+          var v = addon.createDeployPlugin({ name: alias });
+          return v;
+        });
+      })
+      .reduce(function(plugins, arr) {
+        return plugins.concat(arr);
+      }, []);
+  },
+
+  _applyDisabledConfig: function(plugins, disabledMap) {
+    return plugins.filter(function(plugin) {
+      return !disabledMap[plugin.name];
+    }) ;
+  },
+
+  _applyRunOrderConfig: function(plugins, runOrderMap) {
+    var self          = this;
+    var graph         = new DAG();
+    var sortedPlugins = [];
+
+    try {
+      plugins.forEach(function(plugin) {
+        var before = self._asArray(runOrderMap[plugin.name], 'before').concat(self._asArray(plugin, 'runBefore'));
+        var after = self._asArray(runOrderMap[plugin.name], 'after').concat(self._asArray(plugin, 'runAfter'));
+        graph.addEdges(plugin.name, plugin, before, after);
+      });
+
+      graph.topsort(function (vertex) {
+        sortedPlugins.push(vertex.value);
+      });
+    } catch(err) {
+      if (/cycle detected/.test(err)) {
+        throw new SilentError('your ember-cli-deploy plugins have a circular dependency:' + err.message);
+      } else {
+        throw err;
+      }
+    }
+
+    return sortedPlugins;
+  },
+
+  _isDeployPluginPack: function(addon) {
+    return this._addonHasKeyword(addon, 'ember-cli-deploy-plugin-pack');
+  },
+
+  _isValidDeployPlugin: function(addon) {
+    return this._addonHasKeyword(addon, 'ember-cli-deploy-plugin') && this._addonImplementsDeploymentHooks(addon);
+  },
+
+  _pluginNameFromAddonName: function(addon) {
+    var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
+    return addon.name.match(pluginNameRegex)[2];
+  },
+
+  _addonHasKeyword: function(addon, keyword) {
+    var keywords = addon.pkg.keywords;
+    return keywords.indexOf(keyword) > -1;
+  },
+
+  _addonImplementsDeploymentHooks: function(addon) {
+    return addon.createDeployPlugin && typeof addon.createDeployPlugin === 'function';
+  },
+
+  _asArray: function(obj, key) {
+    if (!obj || !obj[key]) {
+      return [];
+    }
+    if (typeof obj[key] === 'string') {
+      return [obj[key]];
+    }
+    return obj[key].slice();
+  },
+
+  _warnIfConfigReferencesUninstalledAddons: function(list, validAddons, configKey) {
+    var addonNames = Object.keys(validAddons);
+
+    var unknownAddons = list.filter(function(name) {
+      return addonNames.indexOf(name) === -1;
+    });
+
+    if (unknownAddons.length) {
+      var message = chalk.yellow('You have referenced the following uninstalled addons in `' + configKey + '`:\n');
```

Comment:
nitpick but

is `uninstalled` the right working here? I assume this is replacing the old
`plugins configuration references plugins which are not available.` message right?

maybe should be `You have referenced the following addons in ... and they're not installed` or something similar